crypto/asn1/i2d_evp.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 1995-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* We need to use some deprecated APIs to support the legacy bits */
+#define OPENSSL_SUPPRESS_DEPRECATED
+
+#include <stdio.h>
+#include "internal/cryptlib.h"
+#include <openssl/evp.h>
+#include <openssl/encoder.h>
+#include <openssl/buffer.h>
+#include <openssl/x509.h>
+#include <openssl/rsa.h>         /* For i2d_RSAPublicKey */
+#include <openssl/dsa.h>         /* For i2d_DSAPublicKey */
+#include <openssl/ec.h>          /* For i2o_ECPublicKey */
+#include "crypto/asn1.h"
+#include "crypto/evp.h"
+
+static int i2d_provided(OSSL_LIB_CTX *libctx,
+                        const EVP_PKEY *a, const char *propq,
+                        int selection, const char *output_structures[],
+                        unsigned char **pp)
+{
+    OSSL_ENCODER_CTX *ctx = NULL;
+    int ret;
+
+    for (ret = -1;
+         ret == -1 && *output_structures != NULL;
+         output_structures++) {
+        /*
+         * The i2d_ calls don't take a boundary length for *pp.  However,
+         * OSSL_ENCODER_CTX_get_num_encoders() needs one, so we make one
+         * up.
+         */
+        size_t len = INT_MAX;
+
+        ctx = OSSL_ENCODER_CTX_new_by_EVP_PKEY(a, selection, "DER",
+                                               *output_structures,
+                                               libctx, propq);
+        if (ctx == NULL)
+            return -1;
+        if (OSSL_ENCODER_to_data(ctx, pp, &len))
+            ret = (int)len;
+        OSSL_ENCODER_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    if (ret == -1)
+        ERR_raise(ERR_LIB_ASN1, ASN1_R_UNSUPPORTED_TYPE);
+    return ret;
+}
+
+int i2d_KeyParams_ex(OSSL_LIB_CTX *libctx, const EVP_PKEY *a, const char *propq,
```

Comment:
Should we be thinking about different names for these functions going forward?

Reply:
To what?

(might I remind you of https://github.com/openssl/web/commit/73e69696a67ccd706dae5b8359bb423febde25aa?)

Reply:
Yes that was one way forward..
A name that doesn't suck is another :)
Something that doesn't make you have to make up names for i and d (internal to der?)
EVP_PKEY_to_binary come to mind..

Reply:
> EVP_PKEY_to_binary

So...  OSSL_ENCODER? :wink:

I'll gladly drop them and replace the *one* use case (in one of our test programs) with an OSSL_ENCODER call.
(which I now remember is added by this PR)

Reply:
... and yes, i2d means internal to der.  It's even documented:

>    **i2d_TYPE()** encodes the structure pointed to by a into DER format. 
>
> ...
> **NOTES**
>    The letters **i** and **d** in **i2d_TYPE** stand for "internal" (that is, an
>    internal C structure) and "DER" respectively.  So **i2d_TYPE** converts
>    from internal to DER.

Reply:
> I'll gladly drop them and replace the one use case (in one of our test programs) with an OSSL_ENCODER call.

Note that this comes with a huge caveat: depending on how providers are loaded and in what library context, `i2d_PrivateKey()` and similar will fail...  and since they take an `EVP_PKEY`, it's not entirely obvious.

Reply:
Dropped